File: src/domain/models/stack_profile.py
"""Stack profile protocol and concrete implementations.

A StackProfile provides stack-specific knowledge for the generation pipeline.
Each profile knows its language toolchain, quality gate commands, source layout,
and how to derive a root package name from a project name.

Supported profiles:
- PythonUvProfile: Python + uv (ruff, mypy, pytest, import-linter/pytestarch)
- GenericProfile: Fallback for unknown stacks, skips stack-specific stages
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from src.domain.models.quality_gate import QualityGate


@runtime_checkable
class StackProfile(Protocol):
    """Strategy interface providing stack-specific knowledge for the generation pipeline."""

    @property
    def stack_id(self) -> str: ...

    @property
    def file_glob(self) -> str: ...

    @property
    def project_manifest(self) -> str: ...

    @property
    def source_layout(self) -> tuple[str, ...]: ...

    @property
    def quality_gate_commands(self) -> dict[QualityGate, list[str]]: ...

    @property
    def quality_gate_display(self) -> str: ...

    @property
    def fitness_available(self) -> bool: ...

    def to_root_package(self, project_name: str) -> str: ...


class PythonUvProfile:
    """Python + uv stack profile with full quality gate pipeline."""

    @property
    def stack_id(self) -> str:
        return "python-uv"

    @property
    def file_glob(self) -> str:
        return "**/*.py"

    @property
    def project_manifest(self) -> str:
        return "pyproject.toml"

    @property
    def source_layout(self) -> tuple[str, ...]:
        return ("src/domain/", "src/application/", "src/infrastructure/")

    @property
    def quality_gate_commands(self) -> dict[QualityGate, list[str]]:
        from src.domain.models.quality_gate import QualityGate

        return {
            QualityGate.LINT: ["uv", "run", "ruff", "check", "."],
            QualityGate.TYPES: ["uv", "run", "mypy", "."],
            QualityGate.TESTS: ["uv", "run", "pytest"],
            QualityGate.FITNESS: ["uv", "run", "pytest", "tests/architecture/"],
        }

    @property
    def quality_gate_display(self) -> str:
        return (
            "## Quality Gates\n"
            "\n"
            "```bash\n"
            "uv run ruff check .              # Lint\n"
            "uv run mypy .                    # Type check\n"
            "uv run pytest                    # Tests\n"
            "```\n"
        )

    @property
    def fitness_available(self) -> bool:
        return True

    def to_root_package(self, project_name: str) -> str:
        return project_name.replace("-", "_")


class GenericProfile:
    """Fallback profile for unknown stacks. Skips stack-specific stages."""

    @property
    def stack_id(self) -> str:
        return "generic"

    @property
    def file_glob(self) -> str:
        return "*"

    @property
    def project_manifest(self) -> str:
        return ""

    @property
    def source_layout(self) -> tuple[str, ...]:
        return ()

    @property
    def quality_gate_commands(self) -> dict[QualityGate, list[str]]:
        return {}

    @property
    def quality_gate_display(self) -> str:
        return ""

    @property
    def fitness_available(self) -> bool:
        return False

    def to_root_package(self, project_name: str) -> str:
        return project_name

File: internal/shared/domain/valueobjects/stack_profile.go
package valueobjects

import "strings"

// StackProfile is the strategy interface providing stack-specific knowledge.
type StackProfile interface {
	StackID() string
	FileGlob() string
	ProjectManifest() string
	SourceLayout() []string
	QualityGateCommands() map[QualityGate][]string
	QualityGateDisplay() string
	FitnessAvailable() bool
	ToRootPackage(projectName string) string
}

// PythonUvProfile provides full Python+uv pipeline values.
type PythonUvProfile struct{}

// StackID returns the stack identifier.
func (p PythonUvProfile) StackID() string { return "python-uv" }

// FileGlob returns the file glob pattern.
func (p PythonUvProfile) FileGlob() string { return "**/*.py" }

// ProjectManifest returns the project manifest filename.
func (p PythonUvProfile) ProjectManifest() string { return "pyproject.toml" }

// SourceLayout returns the DDD source layout directories.
func (p PythonUvProfile) SourceLayout() []string {
	return []string{"src/domain/", "src/application/", "src/infrastructure/"}
}

// QualityGateCommands returns commands for each quality gate.
func (p PythonUvProfile) QualityGateCommands() map[QualityGate][]string {
	return map[QualityGate][]string{
		QualityGateLint:    {"uv", "run", "ruff", "check", "."},
		QualityGateTypes:   {"uv", "run", "mypy", "."},
		QualityGateTests:   {"uv", "run", "pytest"},
		QualityGateFitness: {"uv", "run", "pytest", "tests/architecture/"},
	}
}

// QualityGateDisplay returns the markdown display for quality gates.
func (p PythonUvProfile) QualityGateDisplay() string {
	return "## Quality Gates\n" +
		"\n" +
		"```bash\n" +
		"uv run ruff check .              # Lint\n" +
		"uv run mypy .                    # Type check\n" +
		"uv run pytest                    # Tests\n" +
		"```\n"
}

// FitnessAvailable returns whether fitness tests are available.
func (p PythonUvProfile) FitnessAvailable() bool { return true }

// ToRootPackage converts a project name to a root package name.
func (p PythonUvProfile) ToRootPackage(projectName string) string {
	return strings.ReplaceAll(projectName, "-", "_")
}

// GenericProfile is a fallback profile for unknown stacks.
type GenericProfile struct{}

// StackID returns the stack identifier.
func (g GenericProfile) StackID() string { return "generic" }

// FileGlob returns the file glob pattern.
func (g GenericProfile) FileGlob() string { return "*" }

// ProjectManifest returns the project manifest filename.
func (g GenericProfile) ProjectManifest() string { return "" }

// SourceLayout returns the source layout directories.
func (g GenericProfile) SourceLayout() []string { return nil }

// QualityGateCommands returns commands for each quality gate.
func (g GenericProfile) QualityGateCommands() map[QualityGate][]string { return nil }

// QualityGateDisplay returns the markdown display for quality gates.
func (g GenericProfile) QualityGateDisplay() string { return "" }

// FitnessAvailable returns whether fitness tests are available.
func (g GenericProfile) FitnessAvailable() bool { return false }

// ToRootPackage converts a project name to a root package name.
func (g GenericProfile) ToRootPackage(projectName string) string { return projectName }

// GoModProfile provides full Go modules pipeline values.
type GoModProfile struct{}

// StackID returns the stack identifier.
func (g GoModProfile) StackID() string { return "go-mod" }

// FileGlob returns the file glob pattern.
func (g GoModProfile) FileGlob() string { return "**/*.go" }

// ProjectManifest returns the project manifest filename.
func (g GoModProfile) ProjectManifest() string { return "go.mod" }

// SourceLayout returns the DDD source layout directories.
func (g GoModProfile) SourceLayout() []string {
	return []string{"internal/*/domain/", "internal/*/application/", "internal/*/infrastructure/"}
}

// QualityGateCommands returns commands for each quality gate.
func (g GoModProfile) QualityGateCommands() map[QualityGate][]string {
	return map[QualityGate][]string{
		QualityGateLint:    {"golangci-lint", "run", "./..."},
		QualityGateTypes:   {"go", "vet", "./..."},
		QualityGateTests:   {"go", "test", "-race", "./..."},
		QualityGateFitness: {"arch-go"}, // MIT-licensed architecture testing tool
	}
}

// QualityGateDisplay returns the markdown display for quality gates.
func (g GoModProfile) QualityGateDisplay() string {
	return "## Quality Gates\n" +
		"\n" +
		"```bash\n" +
		"golangci-lint run ./...          # Lint\n" +
		"go vet ./...                     # Type check\n" +
		"go test -race ./...              # Tests\n" +
		"arch-go                          # Architecture fitness\n" +
		"```\n"
}

// FitnessAvailable returns whether fitness tests are available.
func (g GoModProfile) FitnessAvailable() bool { return true }

// ToRootPackage converts a project name to a root package name.
func (g GoModProfile) ToRootPackage(projectName string) string { return projectName }

// Compile-time interface checks.
var (
	_ StackProfile = PythonUvProfile{}
	_ StackProfile = GenericProfile{}
	_ StackProfile = GoModProfile{}
)
